Mathlib/Analysis/Calculus/IteratedDeriv/Lemmas.lean

```diff
@@ -34,6 +34,18 @@ theorem iteratedDerivWithin_congr (hfg : Set.EqOn f g s) :
     rw [iteratedDerivWithin_succ, iteratedDerivWithin_succ]
     exact derivWithin_congr (IH hfg) (IH hfg hy)
 
+theorem iteratedDerivWithin_eq_iteratedDeriv
+(hf : ContDiff 𝕜 (⊤ : ℕ∞) f) (hs : UniqueDiffOn 𝕜 s) (hx : x ∈ s):
+iteratedDerivWithin n f s x = iteratedDeriv n f x := by
+  revert x; induction' n with n hn
+  · simp
+  · intro x hx
+    rw [iteratedDerivWithin_succ, iteratedDeriv_succ, derivWithin, deriv]
+    rw [fderivWithin_congr @hn (@hn x hx)]
+    rw [fderivWithin_eq_fderiv (UniqueDiffOn.uniqueDiffWithinAt hs hx)]
+    apply Differentiable.differentiableAt
+          (ContDiff.differentiable_iteratedDeriv n hf (Batteries.compareOfLessAndEq_eq_lt.mp rfl))
```

Comment:
Here's a bit of golfing. Note that you can use `induction n generalizing x ...` to avoid reverting `x` in your proof.
```suggestion
theorem iteratedDerivWithin_eq_iteratedDeriv
    (hf : ContDiff 𝕜 (⊤ : ℕ∞) f) (hs : UniqueDiffOn 𝕜 s) (hx : x ∈ s) :
    iteratedDerivWithin n f s x = iteratedDeriv n f x := by
  induction n generalizing x with
  | zero => simp
  | succ n hn =>
    rw [iteratedDerivWithin_succ, iteratedDeriv_succ, derivWithin, deriv,
      fderivWithin_congr @hn (@hn x hx), fderivWithin_eq_fderiv (hs.uniqueDiffWithinAt hx)]
    apply (hf.differentiable_iteratedDeriv n
      (ENat.natCast_lt_of_coe_top_le_withTop (le_refl _) _)).differentiableAt
```